Sources/ComposableArchitecture/Instrumentation.swift

```diff
@@ -0,0 +1,142 @@
+import Foundation
+
+/// Interface to enable tracking/instrumenting the activity within TCA as ``Actions`` are sent into ``Store``s and
+/// ``ViewStores``, ``Reducers`` are executed, and ``Effects`` are observed.
+///
+/// The way the library will call the closures provided is identical to the way that the ``Actions`` and ``Effects`` are
+/// handled internally. That means that there is likely to be ``Instrumentation.ViewStore`` `will|did` pairs contained
+/// within the bounds of an ``Instrumentation.Store`` `will|did` pair. For example: Consider sending a simple ``Action``
+/// into a ``ViewStore`` that does not produce any synchronous ``Effects`` to be generated by a ``Reducer``, and the
+/// ``ViewStore`` is scoped off a parent ``Store``s state:
+///   1. An ``Action`` is passed to ``ViewStore.send``
+///   2. The ``Instrumentation.ViewStore.willSend`` callback is called
+///   3. The ``Action`` is passed to ``Store.send``
+///   4. The ``Instrumentation.Store.willSend`` callback is called
+///   5. The ``Instrumentation.Store.willProcessEvents`` callback is called
+///   6. The ``Action`` is passed to the ``Store.reducer`` and acted upon
+///   7. The ``Store.reducer`` returns an ``Effect.none`` producing no further ``Action``s
+///   8. The ``Instrumentation.Store.didProcessEvents`` callback is called
+///   9. The ``Instrumentation.Store.willScope`` callback is called
+///   10. The ``Store.state.value`` is updated
+///   11. The ``ViewStore`` is updated with the ``Store.state``s new value
+///   12. The ``Instrumentation.ViewStore.willDeduplicate`` callback is called
+///   13. The state is deduplicated
+///   14. The ``Instrumentation.ViewStore.didDeduplicate`` callback is called
+///   15. The ``Instrumentation.ViewStore.stateWillChange`` callback is called
+///   16. The ``ViewStore.state`` value is updated
+///   17. All subscribers to the ``ViewStore.state`` are updated with the new state value
+///   18. The ``Instrumentation.ViewStore.stateDidChange`` callback is called
+///   19. The ``Instrumentation.Store.didScope`` callback is called
+///   20. The ``Instrumentation.Store.didSend`` callback is called
+///   21. The ``Instrumentation.ViewStore.didSend`` callback is called
+public class Instrumentation {
+  public typealias Trigger = (EventInfo) -> Void
```

Comment:
IMO s/Trigger/Callback/ or Report or somesuch; the callers (which is who this name is for) consider this an observer, not an actor (so e.g. "Triggered" would be better than "Trigger", but IMO is not a good name, either)